### DISCUSSION TITLE
PyInstaller --onefile flag

### DISCUSSION THREAD

**Discussion by alexx-ftw** [dennisvang/notsotuf-example#36] [Q&A] [answered]
Hello,

Firstly, thanks for this great utility.

I am using PyInstaller with the `--onefile` flag.

What would need to be changed in this example for it to work properly with the `--onefile` setting?

**Comment by dennisvang** [comment, marked as answer]
>What would need to be changed in this example for it to work properly with the --onefile setting?

Hi @alexx-ftw, thanks, good question.

Two things need to be done, both related to pyinstaller itself:

1. The [main.spec] file would need to be changed slightly, removing the `COLLECT()` call and moving the script, binaries, and zipfiles into the `EXE()` call (see [pyinstaller docs]). 

   The result looks like this:
    ```python
    # -*- mode: python ; coding: utf-8 -*-
    
    
    block_cipher = None
    
    
    a = Analysis(
        ['src/main.py'],
        pathex=[],
        binaries=[],
        datas=[('temp_my_app/repository/metadata/root.json', '.')],
        hiddenimports=[],
        hookspath=[],
        hooksconfig={},
        runtime_hooks=[],
        excludes=[],
        win_no_prefer_redirects=False,
        win_private_assemblies=False,
        cipher=block_cipher,
        noarchive=False,
    )
    pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)
    
    exe = EXE(
        pyz,
        a.scripts,
        a.binaries,
        a.datas,
        a.zipfiles,
        [],
        name='main',
        debug=False,
        bootloader_ignore_signals=False,
        strip=False,
        upx=True,
        upx_exclude=[],
        runtime_tmpdir=None,
        console=True,
        disable_windowed_traceback=False,
        argv_emulation=False,
        target_arch=None,
        codesign_identity=None,
        entitlements_file=None,
    )
    ```

2. The [pyinstaller call] needs to be modified to place the resulting executable file in a subdirectory:
    ```python
    pyinstaller.exe "$repo_dir\main.spec" --clean -y --distpath "$temp_dir\dist\myapp" --workpath "$temp_dir\build"
                                                                                ^^^^^
    ``` 
Note that `tufup` does not know about pyinstaller, and does not care. It just moves bundles of files. Any type of files. What matters is that you tell `tufup` where to find those files.


[pyinstaller docs]: https://pyinstaller.org/en/stable/spec-files.html#spec-file-operation
[main.spec]: https://github.com/dennisvang/tufup-example/blob/a284a4a4c4bb415199af08a151b2356856e72e95/main.spec
[pyinstaller call]: https://github.com/dennisvang/tufup-example/blob/a284a4a4c4bb415199af08a151b2356856e72e95/test_update_cycle.ps1#L71

**Reply by alexx-ftw** [reply]
That is great, and changes are very small overall.

Another followup question: would that allow to apply Patches instead of full updates?

**Reply by dennisvang** [reply]
>[...] would that allow to apply Patches instead of full updates?

Patches should work, out-of-the-box, unless you disable them on the *repo* side ([script] or [cli]) and/or *[client]* side. You can try them out in the example.

Also see [test_update_cycle.ps1].

Do note, however, that there's an open issue related to the patch-creation process taking too much time under certain circumstances (difficult to reproduce). See dennisvang/tufup#154

[script]: https://github.com/dennisvang/tufup/blob/473045f2f8551e3bef9febd8b9969851c20df49a/src/tufup/repo/__init__.py#L750
[cli]: https://github.com/dennisvang/tufup/blob/473045f2f8551e3bef9febd8b9969851c20df49a/src/tufup/repo/cli.py#L98
[client]: https://github.com/dennisvang/tufup/blob/473045f2f8551e3bef9febd8b9969851c20df49a/src/tufup/client.py#L148
[test_update_cycle.ps1]:  https://github.com/dennisvang/tufup-example/blob/master/test_update_cycle.ps1
